test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/CpsPackageReferenceProjectTests.cs

```diff
@@ -4127,6 +4126,107 @@ public async Task GetInstalledAndTransitivePackagesAsync_TransitiveOriginsParamS
+            await RestorePackageSpecsAsync(rootDir, output: null, initialProjectSpec);
+
+            CpsPackageReferenceProject project = PrepareCpsRestoredProject(initialProjectSpec, cache);
+
+            // Act
+            var installedPackages = await project.GetInstalledPackagesAsync(CancellationToken.None); // This will change IsInstalledAndTransitiveComputationNeeded to false
+            ProjectPackages result = await project.GetInstalledAndTransitivePackagesAsync(includeTransitivePackages, includeTransitiveOrigins, CancellationToken.None);
+
+            // Assert
+            var transitivePackagesCount = includeTransitivePackages ? 1 : 0;
+            var transitiveOriginsCount = (includeTransitivePackages && includeTransitiveOrigins) ? 1 : 0;
+
+            Assert.Equal(result.InstalledPackages.Count, installedPackages.Count());
+            Assert.Equal(1, result.InstalledPackages.Count);
+            Assert.Equal(transitivePackagesCount, result.TransitivePackages.Count);
+            Assert.All(result.TransitivePackages, pkg => Assert.Equal(transitiveOriginsCount, pkg.TransitiveOrigins.Count()));
+        }
+
+        [Theory]
+        [InlineData(false, false)]
+        [InlineData(false, true)]
+        [InlineData(true, false)]
+        [InlineData(true, true)]
+        public async Task GetInstalledAndTransitivePackagesAsync_IsInstalledAndTransitiveComputationNeeded_True_Succeed(bool includeTransitivePackages, bool includeTransitiveOrigins)
+        {
+            using var rootDir = new SimpleTestPathContext();
+            await CreatePackagesAsync(rootDir, packageAVersion: "1.0.0");
+            IProjectSystemCache cache = new ProjectSystemCache();
+
+            // Arrange
+            PackageSpec initialProjectSpec = ProjectTestHelpers.GetPackageSpec("MyProject", rootDir.SolutionRoot, framework: "net472", dependencyName: "PackageA");
+            await RestorePackageSpecsAsync(rootDir, output: null, initialProjectSpec);
+
+            CpsPackageReferenceProject project = PrepareCpsRestoredProject(initialProjectSpec, cache);
+
+            // Act
+            ProjectPackages result = await project.GetInstalledAndTransitivePackagesAsync(includeTransitivePackages, includeTransitiveOrigins, CancellationToken.None);
+
+            // Assert
+            var transitivePackagesCount = includeTransitivePackages ? 1 : 0;
+            var transitiveOriginsCount = (includeTransitivePackages && includeTransitiveOrigins) ? 1 : 0;
+
+            Assert.Equal(1, result.InstalledPackages.Count);
+            Assert.Equal(transitivePackagesCount, result.TransitivePackages.Count);
+            Assert.All(result.TransitivePackages, pkg => Assert.Equal(transitiveOriginsCount, pkg.TransitiveOrigins.Count()));
+        }
+
+        [Fact]
+        public async Task GetInstalledAndTransitivePackagesAsync_HasExpectedCache_Succeed()
```

Comment:
What does it mean to have expected cache and succeed? Please make sure the test method describes that more clearly.